handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java

```diff
@@ -0,0 +1,203 @@
+
+/**
+ * This handler supports trivial emulation of soTimeout for clients, measure time between
+ * last <code>write()</code> method invocation and first proceed <code>channelRead()</code>
+ *
+ * NB: Timeout measurement in default implementation
+ * will be cancelled just after first potentially partial <code>read()</code> in channel, so,
+ * if you have marker of last event - it's worth to override {@link timeoutShouldBeCancelled} method,
+ * otherwise - use this handler in pair with reasonable {@link IdleStateHandler} to catch situations
+ * of extremely slow or dead reads.
+ *
+ */
+public class ReadAfterWriteTimeoutHandler extends ChannelDuplexHandler {
+    private static final long MIN_TIMEOUT_NANOS = TimeUnit.MILLISECONDS.toNanos(1);
+
+    private final long timeoutNanos;
+    private final boolean closeOnTimeout;
+
+    //shared state, consider removing
+    private ChannelFutureListener channelFutureListener;
+    private ReadTimeoutTask readTimeoutTask;
+
+    private long readFinishTime;
+    private long writeFinishTime;
+    private volatile boolean removed;
+    private boolean closed;
+    private ScheduledFuture<?> timeout;
+
+    /**
+     * Creates a new instance.
+     *
+     * @param timeout
+     *        read timeout after last write to the channel
+     * @param unit
+     *        the {@link TimeUnit} of {@code timeout}
+     * @param closeOnTimeout if channel should be closed by this handler after firing {@link ReadTimeoutException}
+     *
+     */
+    public ReadAfterWriteTimeoutHandler(long timeout, TimeUnit unit, boolean closeOnTimeout) {
+        if (timeout <= 0) {
+            timeoutNanos = 0;
+        } else {
+            timeoutNanos = Math.max(unit.toNanos(timeout), MIN_TIMEOUT_NANOS);
+        }
+        this.closeOnTimeout = closeOnTimeout;
+    }
+
+    /**
+     * Creates a new instance.
+     *
+     * @param timeout
+     *        read timeout
+     * @param unit
+     *        the {@link TimeUnit} of {@code timeout}
+     */
+    public ReadAfterWriteTimeoutHandler(long timeout, TimeUnit unit) {
+        this(timeout, unit, true);
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        init(ctx);
+    }
+
+    @Override
+    public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+        destroy();
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        destroy();
+        super.channelInactive(ctx);
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        readFinishTime = System.nanoTime();
+        if (timeoutShouldBeCancelled(ctx, msg)) {
+            disable();
+        }
+        super.channelRead(ctx, msg);
+    }
+
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
+        ctx.write(msg, promise.addListener(channelFutureListener));
```

Comment:
Yep, you are right. Added ```disable()``` call to the listener before scheduling new timeout. https://github.com/netty/netty/pull/5484/files#diff-6a2c899583a1eae50e09ae517f50121aR94 Thanks for catching. WDYT about this fix?